controller-server/src/main/java/com/yahoo/vespa/hosted/controller/MailVerifier.java

```diff
@@ -0,0 +1,114 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.controller;
+
+import com.yahoo.config.provision.TenantName;
+import com.yahoo.vespa.hosted.controller.api.integration.organization.Mailer;
+import com.yahoo.vespa.hosted.controller.persistence.CuratorDb;
+import com.yahoo.vespa.hosted.controller.tenant.CloudTenant;
+import com.yahoo.vespa.hosted.controller.tenant.Tenant;
+import com.yahoo.vespa.hosted.controller.tenant.TenantContacts;
+import com.yahoo.vespa.hosted.controller.tenant.TenantInfo;
+import com.yahoo.vespa.hosted.controller.tenant.PendingMailVerification;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.UUID;
+
+
+/**
+ * @author olaa
+ */
+public class MailVerifier {
+
+    private final TenantController tenantController;
+    private final Mailer mailer;
+    private final CuratorDb curatorDb;
+    private final Clock clock;
+    private static final Duration VERIFICATION_DEADLINE = Duration.ofDays(7);
+
+    public MailVerifier(TenantController tenantController, Mailer mailer, CuratorDb curatorDb, Clock clock) {
+        this.tenantController = tenantController;
+        this.mailer = mailer;
+        this.curatorDb = curatorDb;
+        this.clock = clock;
+    }
+
+    public PendingMailVerification sendMailVerification(TenantName tenantName, String email, PendingMailVerification.MailType mailType) {
+        var verificationCode = UUID.randomUUID().toString();
+        var verificationDeadline = clock.instant().plus(VERIFICATION_DEADLINE);
+        var pendingMailVerification = new PendingMailVerification(tenantName, email, verificationCode, verificationDeadline, mailType);
+        mailer.sendVerificationMail(pendingMailVerification);
+        writePendingVerification(pendingMailVerification);
```

Comment:
Maybe it's better to write the pending verification before we send the email? That way we won't send something that we then fail to write.